File: system-test/typescript/src/testkit/dsl/port/myShop/given/steps/given-coupon.ts
import type { GivenStage } from '../given-stage.js';
import type { WhenStage } from '../../when/when-stage.js';
import type { ThenStage } from '../../then/then-stage.js';

export interface GivenCoupon {
  withCouponCode(couponCode: string): GivenCoupon;
  withDiscountRate(discountRate: string | number): GivenCoupon;
  withValidFrom(validFrom: string): GivenCoupon;
  withValidTo(validTo: string): GivenCoupon;
  withUsageLimit(usageLimit: string | number): GivenCoupon;
  and(): GivenStage;
  when(): WhenStage;
  then(): ThenStage;
}

File: system/multitier/backend-java/src/testSupport/java/com/mycompany/myshop/backend/support/port/given/steps/GivenCoupon.java
package com.mycompany.myshop.backend.support.port.given.steps;

import com.mycompany.myshop.backend.support.port.given.steps.base.GivenStep;

public interface GivenCoupon extends GivenStep {
    GivenCoupon withCouponCode(String couponCode);

    GivenCoupon withDiscountRate(String discountRate);

    GivenCoupon withDiscountRate(double discountRate);

    GivenCoupon withValidFrom(String validFrom);

    GivenCoupon withValidTo(String validTo);

    GivenCoupon withUsageLimit(int usageLimit);
}
